lib/membrane_webm/parser.ex

```diff
@@ -0,0 +1,276 @@
+defmodule Membrane.WebM.Parser do
+  use Membrane.Filter
+
+  alias Membrane.{Buffer, Time}
+  alias Membrane.WebM.Parser.Vint
+  alias Membrane.WebM.Schema
+
+  def_input_pad :input,
+    availability: :always,
+    mode: :pull,
+    demand_unit: :buffers,
+    caps: :any
+
+  def_output_pad :output,
+    availability: :always,
+    mode: :pull,
+    caps: :any
+
+  @return_elements [
+    :EBML,
+    :SeekHead,
+    :Info,
+    :Tracks,
+    :Tags,
+    :Cues,
+    :Cluster
+  ]
+
+  @impl true
+  def handle_init(_) do
+    {:ok, %{acc: <<>>}}
+  end
+
+  @impl true
+  def handle_demand(:output, size, :buffers, _context, state) do
+    {{:ok, demand: {:input, size}}, state}
+  end
+
+  @impl true
+  def handle_process(:input, %Buffer{payload: payload}, _context, %{acc: acc}) do
+    unparsed = acc <> payload
+
+    case parse_many(unparsed, []) do
+      {:ok, {name, _data} = parsed_element, rest} ->
+        IO.puts("Parser sending #{name}")
+        {{:ok, buffer: {:output, %Buffer{payload: parsed_element}}}, %{acc: rest}}
+
+      :need_more_bytes ->
+        {{:ok, redemand: :output}, %{acc: unparsed}}
+    end
+  end
+
+  def parse_many(bytes, acc) do
+    case parse_element(bytes) do
+      {{name, _data} = element, rest} ->
+        if name in @return_elements do
+          {:ok, element, rest}
+        else
+          if rest == <<>> do
+            [element | acc]
+          else
+            parse_many(rest, [element | acc])
+          end
+        end
+
+      :need_more_bytes ->
+        :need_more_bytes
+
+      {:ok, element, rest} ->
+        {:ok, element, rest}
+    end
+  end
+
+  def parse_element(bytes) do
+    %{vint: vint, rest: bytes} = Vint.parse(bytes)
+    id = vint.element_id
+    %{vint: vint, rest: bytes} = Vint.parse(bytes)
+    # TODO: deal with unknown data size
+    data_size = vint.vint_data
+    name = Schema.element_id_to_name(id)
+    type = Schema.element_type(name)
+
+    if name == :Unknown do
+      IO.warn("unknown element ID: #{id}")
```

Comment:
Use `Membrane.Logger.warn` instead. Just note, never use `IO.inspect`, `IO.puts`, `IO.warn` when you are putting something for review.

Reply:
And should we raise here? Are you able to operate on `:Unknown` element?